csm_web/scheduler/views/section.py

```diff
@@ -779,3 +784,198 @@ def wotd(self, request, pk=None):
+                log_str(receiver.user),
+            )
+            return Response({}, status=status.HTTP_201_CREATED)
+
+    
+    @action(detail=True, methods=["POST", "DELETE"], url_path='swap/(?P<swap_id>[^/.]+)')
+    def swap_with_id(self, request, section_id=None, swap_id=None):
+        '''If swap_id is not None, initiate the swap between the two students. This should 
+        be done in one atomic transaction. If either or both of the swaps are already 
+        deleted, then return a 404 error.'''
+        student_id = request.data.get("student_id", -1) if request.method != "GET" \
+            else request.query_params.get("student_id", -1)
+        if student_id == -1:
+            raise PermissionDenied("The `student_id` field in the request cannot be empty,"
+                                   "please specify student.")
+        try:
+            student = get_object_or_error(Student.objects, id=student_id)
+        except Exception:
+            raise NotFound("No student found with id: " + str(student_id))
+        if request.method == "POST":
+            with transaction.atomic():
+                # Check if swap object with given id exists
+                try:
+                    target_swap = get_object_or_error(Swap.objects.select_for_update(), id=swap_id)
+                except Exception as e:
+                    logger.info(e)
+                    logger.error(
+                        "<Swap Processed:Failure> User %s could not complete swap. "
+                        "Swap does not exist.",
+                        log_str(student.user),
+                    )
+                    raise NotFound("Swap with id: " + str(swap_id) + " does not exist.")
+
+                # Execute atomic transaction, and swap sections
+                try:
+                    sender = get_object_or_error(Student.objects.select_for_update(), 
+                                                 id=target_swap.sender.id)
+                    receiver = get_object_or_error(Student.objects.select_for_update(), 
+                                                   id=target_swap.receiver.id)
+                except Exception as e:
+                    logger.error(
+                        "<Swap Processed:Failure> User %s could not swap with %s",
+                        log_str(student.user),
+                        log_str(receiver.user),
+                    )
+                    raise NotFound("Could not find swaps for student.")
+                sender.section = target_swap.receiver.section
+                # Get outgoing and incoming swaps for sender
+                outgoing_swaps = Swap.objects.filter(sender=sender).select_for_update()
+                incoming_swaps = Swap.objects.filter(receiver=sender).select_for_update()
+                for expired_swap in outgoing_swaps:
+                    expired_swap.delete()
+                for expired_swap in incoming_swaps:
+                    expired_swap.delete()
```

Comment:
There's a bulk delete method for a `QuerySet`: https://docs.djangoproject.com/en/4.2/ref/models/querysets/#delete

Using a for loop sends separate queries for each delete action, while a delete on a query set will send one query to delete all of the objects in the query set.